[FLINK-22757][docs] Adding gcs documentation. Connecting flink to gcs. (#15991)
apache/flink: docs/content/docs/deployment/filesystems/gcs.md

---
title: Google Cloud Storage
weight: 3
type: docs
aliases:
  - /deployment/filesystems/gcs.html
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Google Cloud Storage

> [Google Cloud storage](https://cloud.google.com/storage) (GCS) provides cloud storage for a variety of use cases. You can use it for **reading** and **writing data**.

Reviewer: ```You can use it for **reading** and **writing data** as well in conjunction with the [streaming **state backends**]({{< ref "docs/ops/state/state_backends" >}})```.

> [Google Cloud storage](https://cloud.google.com/storage) (GCS) provides cloud storage for a variety of use cases. You can use it for **reading** and **writing data**.

Reviewer: ```suggestion
[Google Cloud storage](https://cloud.google.com/storage) (GCS) provides cloud storage for a variety of use cases. You can use it for **reading** and **writing data** as well in conjunction with the [streaming **state backends**]({{< ref "docs/ops/state/state_backends" >}}).
```

---

> title: Google cloud storage

Reviewer 2: ```suggestion
title: Google Cloud Storage
```

weight: 3
type: docs
aliases:
  - /deployment/filesystems/gcs.html
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Google Cloud Storage

> [Google Cloud storage](https://cloud.google.com/storage) (GCS) provides cloud storage for variety of usecases. You can use it for **reading** and **writing data**.

Reviewer 2: By its exclusion, this seems to suggest that GCS cannot be used for checkpointing. Is that the case?

```suggestion
[Google Cloud storage](https://cloud.google.com/storage) (GCS) provides cloud storage for a variety of use cases. You can use it for **reading** and **writing data**.
```

You can use GCS objects like regular files by specifying paths in the following format:

```plain
gs://<your-bucket>/<endpoint>
```

The endpoint can either be a single file or a directory, for example:

```java
// Read from GSC bucket
env.readTextFile("gs://<bucket>/<endpoint>");

// Write to GCS bucket
stream.writeAsText("gs://<bucket>/<endpoint>");

```

### Libraries

> You would require to include the following libraries to `/lib` for connecting flink with gcs. 

Reviewer 2: Can the GCS filesystem be used as a Flink plugin, or is it really the case that it only works by putting the jar files into the lib directory (and I assume you meant Flink's lib directory, rather than `/lib`).

```suggestion
You must include the following libraries in Flink's `lib` directory to connect Flink with gcs:
```

Author: Hi, I tried running the same application via copying jar's to `plugins/gcs`. It does not seem to work with that.

> You must include the following libraries in Flink's `lib` directory to connect Flink with gcs:

Reviewer 3: Because the packages below are described in maven format, let's maybe make it clear that we are talking about copying JARs into the lib folder.

Author: Using the keyword `jars` instead of `libraries` in the first line.
`You must include the following jars in Flink's lib`. I hope that makes it more clear

```xml
<dependency>
  <groupId>org.apache.flink</groupId>
  <artifactId>flink-shaded-hadoop2-uber</artifactId>
  <version>${flink.shared_hadoop_latest_version}</version>
</dependency>

> <dependency>

Reviewer 3: Here is the link that tracks the latest version of the connector:
https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-latest-hadoop2.jar

  <groupId>com.google.cloud.bigdataoss</groupId>
  <artifactId>gcs-connector</artifactId>
  <version>hadoop2-2.2.0</version>
</dependency>
```

We have tested with `flink-shared-hadoop2-uber version` >= `2.8.3-1.8.3`.

### Authentication to access GCS

> You would need authentication to access GCS. Please follow the [link](https://cloud.google.com/storage/docs/authentication).

Reviewer 2: ```suggestion
Most operations on GCS require authentication. Please see [the documentation on Google Cloud Storage authentication](https://cloud.google.com/storage/docs/authentication) for more information.
```

> You can expose the key generated via `GOOGLE_APPLICATION_CREDENTIALS` environment variable. 

Reviewer 2: ```suggestion
You can provide the necessary key via the `GOOGLE_APPLICATION_CREDENTIALS` environment variable. 
```
